File: BestTimetoBuyandSellStockwithCooldown.cpp
/*
** Time Limit Exceeded
*/
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
    	
    	return maxProfit(prices, 0);
    	
    }
    
    int maxProfit(vector<int> prices, int begin) {
    	int m = 0;
    	for(int i = begin + 1; i < prices.size(); i++) {
    		if(prices[i] > prices[i - 1]) {
    			m = max(m, prices[i] - prices[begin] + maxProfit(prices, i + 2));
			}else {
				m = max(m, maxProfit(prices, i));
			}
		}
		return m;
	}
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> prices;
	for(int i = 0; i < n; i++) {
		int price;
		cin>>price;
		prices.push_back(price);
	}
	
	Solution *solution = new Solution();
	cout<<solution->maxProfit(prices);
	
	return 0;
	
}

File: BestTimetoBuyandSellStockwithCooldown.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        if n == 0:
            return 0
        buy, sell, rest = [0] * n, [0] * n, [0] * n
        for i in range(n):
            if i == 0:
                buy[i] = -prices[i]
            else:
                buy[i] = max(rest[i - 1] - prices[i], buy[i - 1])
                sell[i] = max(buy[i - 1] + prices[i], sell[i - 1])
                rest[i] = max(sell[i - 1], rest[i - 1])
        return sell[-1]


if __name__ == '__main__':
    prices = [int(price) for price in input().split()]
    solution = Solution()
    print(solution.maxProfit(prices))
